server/athenian/api/internal/miners/github/deployment.py

```diff
@@ -2022,3 +2024,132 @@ async def load_jira_issues_for_deployments(deployments: pd.DataFrame,
+    if not anomal_deploys:
+        return
+
+    log.info("Clearing %d anomal first deployments", len(anomal_deploys))
+
+    stmts = []
+
+    tables = (GitHubCommitDeployment, GitHubPullRequestDeployment, GitHubReleaseDeployment)
+    for Table in tables:
+        where_clause = sa.or_(*(
+            sa.and_(
+                Table.acc_id == account,
+                Table.repository_full_name == deploy.repository_full_name,
+                Table.deployment_name == deploy.deployment_name,
+            ) for deploy in anomal_deploys
+        ))
+        stmts.append(sa.delete(Table).where(where_clause))
+
+    # clear GitHubDeploymentFacts by removing prs and commits from facts data
+    depl_facts_stmt = sa.select(GitHubDeploymentFacts).where(
+        sa.and_(
+            GitHubDeploymentFacts.acc_id == account,
+            GitHubDeploymentFacts.deployment_name.in_(d.deployment_name for d in anomal_deploys),
+        ),
+    )
+    depl_facts_rows = await pdb.fetch_all(depl_facts_stmt)
+    for row in depl_facts_rows:
+        facts = DeploymentFacts(
+            row[GitHubDeploymentFacts.data.name],
+            name=row[GitHubDeploymentFacts.deployment_name.name],
+        )
+        new_facts = facts.with_no_prs_commits()
+        update_stmt = sa.update(GitHubDeploymentFacts).where(
+            sa.and_(*(
+                getattr(GitHubDeploymentFacts, f) == row[f]
+                for f in ("acc_id", "deployment_name", "release_matches", "format_version")
+            )),
+        ).values({GitHubDeploymentFacts.data: new_facts.data})
+        stmts.append(update_stmt)
+
+    await gather(*[pdb.execute(stmt) for stmt in stmts], op="hide_first_deployments")
+
+
+@dataclass(frozen=True, slots=True)
+class _AnomalDeployment:
+    repository_full_name: str
+    deployment_name: str
+
+
+async def _search_anomal_first_deployments(
+    meta_ids: Sequence[int],
+    deployment_facts: pd.DataFrame,
+    mdb: Database,
+    threshold: float = 100.0,
+) -> Sequence[_AnomalDeployment]:
+    """Search the anomal first deployments.
+
+    A deployment is the anomal first for a repo if:
+    - well, it's the first deployment
+    - the time distance with the second deployment is at least `threshold` times
+      higher than the time distance with the repository creation
+
+    Return a sequence of _AnomalDeployment
+
+    """
+    grouped = deployment_facts.explode("repositories").rename(
+        columns={"repositories": "repository"},
+    ).groupby(["environment", "repository"])
+
+    repo_creation_times = {}    # cache repo name => creation time
+
+    result = set()
+
+    log = logging.getLogger(f"{__name__}.hide_first_deployments")
+    for group_name, group in grouped:
+        env, repo = group_name
+        if len(group) == 1:
+            # just one deployment for environment/repo, its prs will not be hidden
+            continue
+        sorted_group = group.sort_values(["started_at"], axis="index")
```

Comment:
Algorithm: compare to the median of the deployment ages. This is more robust because the second deployment alone can be crazy.

Reply:
done